src/components/BarChartCard/BarChartCard.jsx

```diff
@@ -26,54 +26,108 @@ const { iotPrefix } = settings;
-            : groupedDatasets[dataset].map(d => d[series.dataSourceId]),
+      const uniqueGroupNames = [...new Set(values.map(val => val[series.groupDataSourceId]))];
+      const groupedData = uniqueGroupNames.map(group =>
+        values.filter(val => val[series.groupDataSourceId] === group)
+      );
+
+      groupedData.forEach(group => {
+        group.forEach(value => {
+          data.push({
+            group: value[series.labelDataSourceId], // bar
+            value: value[series.dataSourceId], // value
+            key: value[series.groupDataSourceId],
+            ...(series.timeDataSourceId
+              ? { date: new Date(value[series.timeDataSourceId]) } // timestamp
+              : null),
+          });
         });
       });
     } else if (series.labelDataSourceId) {
       // single bars
-      labels = values.map(v => v[series.labelDataSourceId]);
-      datasets = [
-        {
-          ...(series.colors ? { fillColors: series.colors } : {}),
-          data: values.map(v => v[series.dataSourceId]),
-        },
-      ];
+      const uniqueDatasetNames = [...new Set(values.map(val => val[series.labelDataSourceId]))];
+      const labeledData = uniqueDatasetNames.map(name =>
+        values.filter(val => val[series.labelDataSourceId] === name)
+      );
+
+      labeledData.forEach(dataset => {
+        dataset.forEach(value => {
+          data.push({
+            group: value[series.labelDataSourceId], // bar
+            value: value[series.dataSourceId], // value
+          });
+        });
+      });
     } else if (series.timeDataSourceId) {
       // timestamp
-      labels = values.map(v => v[series.timeDataSourceId]);
-      datasets = [
-        {
-          ...(series.colors ? { fillColors: series.colors } : {}),
-          data: values.map(v => ({
-            date: new Date(v[series.timeDataSourceId]),
-            value: v[series.dataSourceId],
-          })),
-        },
-      ];
+      const uniqueDatasetNames = [...new Set(values.map(val => val[series.timeDataSourceId]))];
+      const labeledData = uniqueDatasetNames.map(name =>
+        values.filter(val => val[series.timeDataSourceId] === name)
+      );
+      labeledData.forEach(dataset => {
+        dataset.forEach(value => {
+          const dataDate = new Date(value[series.timeDataSourceId]);
+          data.push({
+            group: capitalize(series.dataSourceId), // bar
+            value: value[series.dataSourceId], // value
+            date: dataDate, // timestamp
+          });
+        });
+      });
     }
   }
 
+  return data;
+};
+
+/**
+ * Maps values to left and bottom axes based on whether layout is vertical
+ * or horizontal, and if the series is grouped or time-based
+ *
+ * @param {Object} series the definition of the plotted series
+ * @param {string} layout vertical or horizontal
+ *
+ * @returns {object} { bottomAxesMapsTo: string, leftAxesMapsTo: string }
+ */
+export const mapValuesToAxes = (series, layout) => {
```

Comment:
can we get a testcase around this function?

Reply:
Added